File: app/geometry.py
"""平面几何：凸包与载荷圆盘（带非负不确定半径）的包含判定。

约定：
- 支承多边形由已部署支腿平面坐标的凸包给出，凸包顶点按逆时针排列。
- 圆盘完整位于凸多边形内，当且仅当圆心到每一条支承边所在直线的
  有符号距离（内侧为正）都不小于半径；距离恰好等于半径（边界相切）
  视为安全。
"""

from __future__ import annotations

import math
from typing import List, Optional, Sequence, Tuple

EPS = 1e-9

Point = Tuple[float, float]


def convex_hull(points: Sequence[Point]) -> List[Point]:
    """返回点集的凸包（Andrew 单调链，逆时针，剔除共线中点）。

    点数不足或全部共线时返回长度小于 3 的列表。
    """
    pts = sorted({(float(x), float(y)) for x, y in points})
    if len(pts) <= 2:
        return pts

    def cross(o: Point, a: Point, b: Point) -> float:
        return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])

    lower: List[Point] = []
    for p in pts:
        while len(lower) >= 2 and cross(lower[-2], lower[-1], p) <= EPS:
            lower.pop()
        lower.append(p)

    upper: List[Point] = []
    for p in reversed(pts):
        while len(upper) >= 2 and cross(upper[-2], upper[-1], p) <= EPS:
            upper.pop()
        upper.append(p)

    return lower[:-1] + upper[:-1]


def min_signed_edge_distance(center: Point, hull: Sequence[Point]) -> float:
    """圆心到凸包各边所在直线的最小有符号距离（逆时针多边形内侧为正）。"""
    cx, cy = center
    distances: List[float] = []
    n = len(hull)
    for i in range(n):
        ax, ay = hull[i]
        bx, by = hull[(i + 1) % n]
        ex, ey = bx - ax, by - ay
        length = math.hypot(ex, ey)
        if length < EPS:
            continue
        # 叉积 ex*(cy-ay) - ey*(cx-ax)，逆时针内侧为正
        signed = (ex * (cy - ay) - ey * (cx - ax)) / length
        distances.append(signed)
    return min(distances)


class SupportEvaluation:
    """当前部署支腿对载荷圆盘的支承评估结果。"""

    __slots__ = ("hull", "safe", "reason", "min_margin")

    def __init__(
        self,
        hull: Optional[List[Point]],
        safe: bool,
        reason: Optional[str] = None,
        min_margin: Optional[float] = None,
    ) -> None:
        self.hull = hull
        self.safe = safe
        self.reason = reason
        # min_margin = 最小边距 - 半径；>=0 即圆盘完整包含
        self.min_margin = min_margin


def evaluate_support(
    deployed_points: Sequence[Point],
    center: Point,
    radius: float,
) -> SupportEvaluation:
    """评估部署支腿凸包是否完整包含载荷圆盘。"""
    n_deployed = len(deployed_points)
    if n_deployed < 3:
        return SupportEvaluation(
            hull=None,
            safe=False,
            reason=f"仅有 {n_deployed} 只部署支腿，至少需要 3 只才能形成支承多边形",
        )

    hull = convex_hull(deployed_points)
    if len(hull) < 3:
        return SupportEvaluation(
            hull=hull if hull else None,
            safe=False,
            reason="部署支腿位置共线（或重合），无法形成有效的支承多边形",
        )

    min_distance = min_signed_edge_distance(center, hull)
    margin = min_distance - radius

    if min_distance < -EPS:
        return SupportEvaluation(
            hull=hull,
            safe=False,
            reason=(
                f"载荷圆心已越出支承面（圆心至支承边最小有符号距离 "
                f"{min_distance:.6f} < 0）"
            ),
            min_margin=margin,
        )

    if margin < -EPS:
        return SupportEvaluation(
            hull=hull,
            safe=False,
            reason=(
                f"载荷圆盘越出支承面：半径 {radius:.6f} 大于圆心至支承边的"
                f"最小距离 {min_distance:.6f}（余量 {margin:.6f}）"
            ),
            min_margin=margin,
        )

    return SupportEvaluation(hull=hull, safe=True, min_margin=margin)

File: src/geometry.js
'use strict';

/**
 * 平面支承几何：凸包、圆盘-凸多边形包含、支承安全评估。
 * 约定：凸包顶点按逆时针（CCW）返回；边界相切视为安全。
 */

const EPS = 1e-9;

function isFiniteNumber(v) {
  return typeof v === 'number' && Number.isFinite(v);
}

function cross2(o, a, b) {
  return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x);
}

/**
 * 单调链凸包。输入 [{x, y}, ...]，返回去重后的 CCW 顶点（不重复首点）。
 * 少于 3 个有效点或全部共线时返回退化的点列（长度 < 3 或面积为 0，由调用方判定）。
 */
function convexHull(points) {
  const sorted = points
    .map((p) => ({ x: p.x, y: p.y }))
    .sort((a, b) => a.x - b.x || a.y - b.y);
  const unique = [];
  for (const p of sorted) {
    const last = unique[unique.length - 1];
    if (!last || last.x !== p.x || last.y !== p.y) unique.push(p);
  }
  if (unique.length <= 2) return unique;

  const lower = [];
  for (const p of unique) {
    while (lower.length >= 2 && cross2(lower[lower.length - 2], lower[lower.length - 1], p) <= 0) {
      lower.pop();
    }
    lower.push(p);
  }
  const upper = [];
  for (let i = unique.length - 1; i >= 0; i -= 1) {
    const p = unique[i];
    while (upper.length >= 2 && cross2(upper[upper.length - 2], upper[upper.length - 1], p) <= 0) {
      upper.pop();
    }
    upper.push(p);
  }
  lower.pop();
  upper.pop();
  return lower.concat(upper);
}

/** 有向面积（CCW 为正）。 */
function polygonArea(poly) {
  let s = 0;
  for (let i = 0; i < poly.length; i += 1) {
    const a = poly[i];
    const b = poly[(i + 1) % poly.length];
    s += a.x * b.y - b.x * a.y;
  }
  return s / 2;
}

/** 点 c 到 CCW 边 a->b 内侧的有向距离（内侧为正）。 */
function edgeMargin(a, b, c) {
  const dx = b.x - a.x;
  const dy = b.y - a.y;
  const len = Math.hypot(dx, dy);
  if (len === 0) return -Infinity;
  return (dx * (c.y - a.y) - dy * (c.x - a.x)) / len;
}

function scaleOf(values) {
  let m = 1;
  for (const v of values) m = Math.max(m, Math.abs(v));
  return m;
}

/**
 * 圆盘 (c, r) 是否完整落在 CCW 凸多边形内（含边界，相切为安全）。
 */
function diskInsidePolygon(poly, c, r) {
  const coords = [c.x, c.y, r];
  for (const p of poly) coords.push(p.x, p.y);
  const tol = EPS * scaleOf(coords);
  for (let i = 0; i < poly.length; i += 1) {
    const a = poly[i];
    const b = poly[(i + 1) % poly.length];
    if (edgeMargin(a, b, c) < r - tol) return false;
  }
  return true;
}

/**
 * 支承安全评估：所有已部署支腿的凸包必须完整包含当前载荷圆盘。
 * 返回 { safe, hull, reason }；reason 为拒绝/不安全的中文说明（安全时为 null）。
 */
function evaluateSupport(legs, posture) {
  const deployed = legs.filter((l) => l.deployed);
  if (deployed.length < 3) {
    return {
      safe: false,
      hull: convexHull(deployed),
      reason: `已部署支腿仅 ${deployed.length} 只，不足 3 只，无法构成安全支承多边形`,
    };
  }
  const hull = convexHull(deployed);
  const coords = [];
  for (const p of hull) coords.push(p.x, p.y);
  const degenerate = hull.length < 3 || Math.abs(polygonArea(hull)) <= EPS * scaleOf(coords) ** 2;
  if (degenerate) {
    return {
      safe: false,
      hull,
      reason: '已部署支腿共线或重合，支承多边形退化，无法形成安全支承',
    };
  }
  if (!diskInsidePolygon(hull, posture, posture.r)) {
    return {
      safe: false,
      hull,
      reason: '载荷投影圆盘将越出支承多边形边界',
    };
  }
  return { safe: true, hull, reason: null };
}

module.exports = {
  EPS,
  isFiniteNumber,
  convexHull,
  polygonArea,
  edgeMargin,
  diskInsidePolygon,
  evaluateSupport,
};
